### ISSUE TITLE
A custom partitioner is needed

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#8672]
### Description of the enhancement or error report
We'd like to have a partitioner to have an input syntax like:
```
[Mesh]
  ...
  [Partitioner]
    type = BlockPartitioner
    blocks = '1 2 3; 4 5; 6'
  [../]
[]
```
This partitioner will try to partition three subdomains described by `blocks` parameter with semi-colons separately. In this case the three subdomains are one with block 1, 2 and 3, another with block 4 and 5 and the third with block 6. When the partitioning of all subdomains are done, the partitioning will be preserved for the entire domain.

### Rationale for the enhancement or information for reproducing the error
We want this capability because we see some significant load imbalance in some of our simulations. Some of the blocks have costly linear auxiliary kernel evaluations, while the rest does not. This custom partitioner will not be optimal on minimizing the communication but we hope the increased communication cost is insignificant than the cost of the imbalanced aux kernel evaluations.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Better parallel efficiency for certain problems.

**Comment by permcody** [comment]
Are you or Sebastian planning to work on this?

**Comment by YaqiWang** [comment]
No. Will this be more efficiently implemented by @jwpeterson ?

**Comment by jwpeterson** [comment]
This could be done pretty easily at the libmesh level.  Is that what you want to do, and then access it through the existing LibmeshPartitioner class? Is this needed right away?

**Comment by YaqiWang** [comment]
That should be fine unless I have control on how to set up the subdomains. Yes, we want this asap. @fdkong is helping us on the preconditioning and he noticed this. If we can fix this, we will get better parallel efficiency on the transient problem for the latest tiger team meeting.

**Comment by YaqiWang** [comment]
We also need a PETSc update for the FSP. But we will need that small bug fix in BoomerAMG pushed into upstream HYPRE.

**Comment by fdkong** [comment]
> We also need a PETSc update for the FSP. But we will need that small bug fix in BoomerAMG pushed into upstream HYPRE.

The small bug fix did not  in `upstream HYPRE` yet?  I am fine with PETSc update. The latest one  also has some bug fixes for the` superlu_dist`. 

But we have already updated PETSc several times recently, and this keeps @milljm busy. Not sure @milljm has time to try again or not?

**Comment by milljm** [comment]
Piece of cake. Just let me know what version of PETSc to grab... and I will implement it in the package.

**Comment by fdkong** [comment]
> Piece of cake. Just let me know what version of PETSc to grab... and I will implement it in the package.

Great! Let me take a test one my own local desktop. Once the test is passed, we could continue proceeding this.

**Comment by YaqiWang** [comment]
I discussed with Rob at HYPRE team here https://github.com/LLNL/hypre/pull/2. They have not made the fix public available. Is it possible that we temporarily apply the patch on our end?

**Comment by milljm** [comment]
We can, sort of. I would need to serve the patched version on say our mooseframework.org site.

**Comment by YaqiWang** [comment]
We can create a LLNL/hypre fork on github for us and we configure PETSc to download the forked hypre instead of its upstream. We can then apply our patches freely in the hypre fork. @fdkong told me he would be trying this out.

**Comment by fdkong** [comment]
@YaqiWang, is it the only thing you want to change, right? https://github.com/fdkong/hypre/commit/d041dee5206e29c738bc25903c79f0f24306dfee

**Comment by YaqiWang** [comment]
Yes.

**Comment by fdkong** [comment]
@milljm, the latest petsc release, with all current patches,   can be downloaded using git:

`    git clone -b maint https://bitbucket.org/petsc/petsc petsc`


And install hypre with new patches using (@fdkong's repo):

`--download-hypre=git://https://github.com/fdkong/hypre.git  --download-hypre-commit=HEAD   `

The other part of configuration should be the same as you have done before.

**Comment by milljm** [comment]
Ref:  https://github.com/idaholab/package_builder/pull/83

**Comment by milljm** [comment]
@YaqiWang  @fdkong Okay. This package has been uploaded and made available (for Darwin machines only for now) at: http://mooseframework.org/getting-started/osx/

In order to use this 'bleeding edge' version of PETSc, you will need to:

```bash
module purge
module load advanced_modules mpich-clang petsc-bleedingedge
```

**Comment by jwpeterson** [comment]
The new partitioner was in the libmesh update that was merged in b95d9a6. I will work with @fdkong to make a MOOSE interface for it.

**Comment by permcody** [comment]
Ticket closed by referenced PRs
